cpp/src/io/avro/reader_impl.cu

```diff
@@ -0,0 +1,502 @@
+   *
+   * @param[in] use_names List of column names to select
+   *
+   * @return List of column names
+   **/
+  auto select_columns(std::vector<std::string> use_names) {
+    std::vector<std::pair<int, std::string>> selection;
+
+    const auto num_avro_columns = static_cast<int>(columns.size());
+    if (!use_names.empty()) {
+      int index = 0;
+      for (const auto &use_name : use_names) {
+        for (int i = 0; i < num_avro_columns; ++i, ++index) {
+          if (index >= num_avro_columns) {
+            index = 0;
+          }
+          if (columns[index].name == use_name &&
+              type_id::EMPTY !=
+                  to_type_id(&schema[columns[index].schema_data_idx])) {
+            selection.emplace_back(index, columns[index].name);
+            index++;
+            break;
+          }
+        }
+      }
+    } else {
+      // Iterate backwards as fastavro returns from last-to-first?!
+      for (int i = num_avro_columns - 1; i >= 0; --i) {
+        auto col_type = to_type_id(&schema[columns[i].schema_data_idx]);
+        CUDF_EXPECTS(col_type != type_id::EMPTY, "Unsupported data type");
+        selection.emplace_back(i, columns[i].name);
+      }
+    }
+    CUDF_EXPECTS(selection.size() > 0, "Filtered out all columns");
+
+    return selection;
+  }
+
+ private:
+  datasource *const source;
+};
+
+rmm::device_buffer reader::impl::decompress_data(
+    const rmm::device_buffer &comp_block_data, cudaStream_t stream) {
+  size_t uncompressed_data_size = 0;
+  hostdevice_vector<gpu_inflate_input_s> inflate_in(
+      _metadata->block_list.size());
+  hostdevice_vector<gpu_inflate_status_s> inflate_out(
+      _metadata->block_list.size());
+
+  if (_metadata->codec == "deflate") {
+    // Guess an initial maximum uncompressed block size
+    uint32_t initial_blk_len = (_metadata->max_block_size * 2 + 0xfff) & ~0xfff;
+    uncompressed_data_size = initial_blk_len * _metadata->block_list.size();
+    for (size_t i = 0; i < inflate_in.size(); ++i) {
+      inflate_in[i].dstSize = initial_blk_len;
+    }
+  } else if (_metadata->codec == "snappy") {
+    // Extract the uncompressed length from the snappy stream
+    for (size_t i = 0; i < _metadata->block_list.size(); i++) {
+      const auto buffer =
+          _source->get_buffer(_metadata->block_list[i].offset, 4);
+      const uint8_t *blk = buffer->data();
+      uint32_t blk_len = blk[0];
+      if (blk_len > 0x7f) {
+        blk_len = (blk_len & 0x7f) | (blk[1] << 7);
+        if (blk_len > 0x3fff) {
+          blk_len = (blk_len & 0x3fff) | (blk[2] << 14);
+          if (blk_len > 0x1fffff) {
+            blk_len = (blk_len & 0x1fffff) | (blk[3] << 21);
+          }
+        }
+      }
+      inflate_in[i].dstSize = blk_len;
+      uncompressed_data_size += blk_len;
+    }
+  } else {
+    CUDF_FAIL("Unsupported compression codec\n");
+  }
+
+  rmm::device_buffer decomp_block_data(uncompressed_data_size);
```

Comment:
Should a `device_memory_resource` be passed into this API to allow customizing how this `device_buffer` is allocated?

Reply:
Yup, good catch. I was also missing some areas for passing the stream in the Avro reader too.